oxide-auth-async/src/code_grant.rs

```diff
@@ -111,21 +111,21 @@ pub mod access_token {
         registrar::RegistrarError,
     };
 
-    #[async_trait(?Send)]
+    #[async_trait]
     pub trait Extension {
         /// Inspect the request and extension data to produce extension data.
         ///
         /// The input data comes from the extension data produced in the handling of the
         /// authorization code request.
         async fn extend(
-            &mut self, request: &dyn Request, data: Extensions,
+            &mut self, request: &(dyn Request + Sync), data: Extensions,
```

Comment:
It seems that taking a `&mut Request` forces us to have `+ Sync` which is logic and fine by me. What I don't understand, is why suddenly the `Cow<'a, dyn QueryParameter + 'static>` in both `WrappedRequest` (in `authorization` and `access_token`) asks for `Clone` trait. And we can't add as a bound as it's not an "auto trait".

Reply:
This line https://github.com/Geobert/oxide-auth/blob/dfe475570e4cceb535776dd7a3deaf85e675243a/oxide-auth-async/src/endpoint/authorization.rs#L145

has this error:
```
error[E0277]: `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
   --> oxide-auth-async\src\endpoint\authorization.rs:145:52
    |
145 |             authorization_code(&mut self.endpoint, &WrappedRequest::new(&mut request)).await;
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
    |
    = help: within `endpoint::authorization::WrappedRequest<'_, R>`, the trait `std::marker::Sync` is not implemented for `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because it appears within the type `&(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because it appears within the type `std::borrow::Cow<'_, (dyn oxide_auth::endpoint::query::QueryParameter + 'static)>`
    = note: required because it appears within the type `endpoint::authorization::WrappedRequest<'_, R>`
    = note: required for the cast to the object type `dyn oxide_auth::code_grant::authorization::Request + std::marker::Sync`
```

Which I happily want to solve by adding `+ Sync` here https://github.com/Geobert/oxide-auth/blob/dfe475570e4cceb535776dd7a3deaf85e675243a/oxide-auth-async/src/endpoint/authorization.rs#L35

And then, suddenly, `Cow` asks for: 
```
error[E0277]: the trait bound `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static): std::clone::Clone` is not satisfied
  --> oxide-auth-async\src\endpoint\authorization.rs:35:5
   |
35 |     query: Cow<'a, dyn QueryParameter + Sync + 'static>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)`
   |
   = note: required because of the requirements on the impl of `std::borrow::ToOwned` for `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)`
   = note: required because it appears within the type `std::borrow::Cow<'a, (dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)>`
   = note: only the last field of a struct may have a dynamically sized type
```

Reply:
Here is why, dumbby me: https://users.rust-lang.org/t/why-adding-sync-bound-asks-for-clone/46215/12

Reply:
I'm okay with having the async primitives `Send` but I _really_ do not understand why this should influence the sync implementation in any way, and it shouldn't either. It seems that this comes from `&dyn Request` being taken as a parameter to `Extension`. This borrows it, which implies it must be `Sync` for the containing future to be `Send`.

One possible fix would be architectural: Don't use `&dyn Request` as a parameter, instead normalize it to a specific type of Request that lives in `oxide-auth-async`, a request type that does _not_ borrow the original request and implements `Sync` and `Send`. For example, that owns the `query` after having cloned it.

Reply:
OK :)

Reply:
> One possible fix would be architectural: Don't use `&dyn Request` as a parameter in the async traits, instead normalize it to a specific type of Request that lives in `oxide-auth-async`, a request type that does _not_ borrow the original request and implements `Sync` and `Send`. For example, that owns the `query` after having cloned it.

That's basically `WrappedRequest` but without `Cow`, isn't it? We already have a mirror of sync WrappedRequest in async, but with the exact same fields, I'll modify them to clone the needed parts of the Request.

Reply:
More or less. I believe that could work :)